test/e2e/framework/util.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2016 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/client-go/1.5/kubernetes"
+	"k8s.io/client-go/1.5/pkg/api"
+	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/1.5/pkg/labels"
+	"k8s.io/client-go/1.5/pkg/util/wait"
+	"k8s.io/client-go/1.5/pkg/util/yaml"
+
+	"github.com/coreos/prometheus-operator/pkg/spec"
+)
+
+func CreateDeployment(kclient kubernetes.Interface, ns string, deploy *v1beta1.Deployment) error {
+	if _, err := kclient.Extensions().Deployments(ns).Create(deploy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func WaitForPodsReady(kclient kubernetes.Interface, ns string, timeout time.Duration, expectedRunning int, opts api.ListOptions) (*v1.PodList, error) {
+	t := time.After(timeout)
+	tick := time.Tick(500 * time.Millisecond)
+
+	for {
+		select {
+		case <-t:
+			return nil, fmt.Errorf("timed out while waiting for pod to be ready")
+		case <-tick:
+			pl, err := kclient.Core().Pods(ns).List(opts)
+			if err != nil {
+				return nil, err
+			}
+
+			running := 0
+			if len(pl.Items) > 0 {
+				for _, p := range pl.Items {
+					if p.Status.Phase == v1.PodRunning {
+						running++
+					}
+				}
+				if running == expectedRunning {
+					return pl, nil
+				}
+			}
+		}
+	}
+}
+
+func PrometheusListOptions(name string) api.ListOptions {
+	return api.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			"app":        "prometheus",
+			"prometheus": name,
+		}),
+	}
+}
+
+func WaitPrometheusTPRReady(httpClient *http.Client, interval, timeout time.Duration, host, ns string) error {
+	return wait.Poll(interval, timeout, func() (bool, error) {
+		resp, err := ListPrometheuses(host, ns, httpClient)
+		if err != nil {
+			return false, err
+		}
+		defer resp.Body.Close()
+
+		switch resp.StatusCode {
+		case http.StatusOK:
+			return true, nil
+		case http.StatusNotFound: // not set up yet. wait.
+			return false, nil
+		default:
+			return false, fmt.Errorf("invalid status code: %v", resp.Status)
+		}
+	})
+}
+
+func ListPrometheuses(host, ns string, httpClient *http.Client) (*http.Response, error) {
+	return httpClient.Get(fmt.Sprintf("%s/apis/monitoring.coreos.com/v1alpha1/namespaces/%s/prometheuses", host, ns))
+}
+
+func MakePromCluster(name string, replicas int) *spec.Prometheus {
```

Comment:
Actually, all advanced tests will probably want to set something beyond number of replicas. Having it as the single argument in this constructor is somewhat arbitrary.

Realistically, all instantiated setups are probably custom and it will be easiest to just specify the structs directly in the test case. As the TPRs are relatively shallow, that will probably still be quite readable and easier to read.